### ISSUE TITLE
Serialization errors when include level is too low

### ISSUE THREAD

**Issue by andersekdahl** [contentful/contentful.net#214]
Serialization correctly picks up related entries and loads them through the content type resolver, even with interfaces. But if content at the last depth contains links to entries and they're interfaces rather than concrete classes the serialization throws a Json.NET exception that it cannot deserialize. Since the link at that level just contains an id and not any more info about the linked entry.

In some cases the solution is just to increase the include parameter, but since content can change without a developer being involved we can get this exception just by an editor moving things around.

I think this needs to be caught/handled in some way by the SDK. Neither the SDK or the custom content type resolver can really handle this case automatically since the link only contains an id, which isn't enough to do anything useful with. Perhaps the content type resolver should be asked something like "I need to create an object of this interface type and all I have is this id, what should I do?". In our case we could have a special implementation of the interface(s) with just a SystemProperties property and handle the case when we run into such an object.

**Comment by Roblinde** [comment]
Yes, I've noticed this myself and I agree it should be handled more gracefully. Not sure what would be the preference. Simply returning `null`, throwing a more excplicit exception, any other ideas?

**Comment by andersekdahl** [comment]
Throwing an exception wouldn't work for us at least, since most of the response is still valid and in some cases we don't even care about data that deep in the tree.

Returning null would certainly be better than the current situation, but it means that it's impossible to recover from. Best case is just missing content for the end user, worst case is that there's some code that assumes that it's not null and we still get an error.

Either way, I'd be happy with null if involving the content type resolver isn't an option.

**Comment by Roblinde** [comment]
Thanks for the input, I'll test a few scenarios and see what we can do.

**Comment by andersekdahl** [comment]
Btw, another scenario I ended up in that's related to this. I've created a custom json converter (that I might not need now with the fixes you're working on/worked on) that tries to find the correct .NET type. But I ended up in a situation where the only thing my converter got was:
```
{
  "sys": {
    "type": "Link",
    "linkType": "Entry",
    "id": "4Cqn9B7T7qhyKG5I8puyAl"
  }
}
```
So I needed to upwards in the object to find `included.Entries` with the JsonReader but I couldn't because it stopped at the parent entry level. I'm assuming that it's because of this:
https://github.com/contentful/contentful.net/blob/master/Contentful.Core/ContentfulClient.cs#L211

Would it be possible to still expose the root document in some way?

**Comment by Roblinde** [comment]
I'm afraid that's a limitation of the json converters, they only get the part of the document they're supposed to deserialize. This is part of the reason why I've had to implement the selective resolving of properties. It's a bit of a pain, but it seems to be for performance reasons. I hope you won't need a custom one once we get this fixed though.

**Comment by Roblinde** [comment]
Ok, so here's my proposed solution. There's already handling for when an editor publishes and entry with a link to an unpublished entry. It will be included in the Errors array of the response. I blogged about it a while ago here: https://robertlinde.se/posts/the-elusive-contentful-error-array/

The problem is that this only applies if the entry is in the errors array, and in our case it's not. I'm thinking we solve this by handling it the same way as an unpublished entry. The case in point being that we _can't_ really resolve it to anything useful as we don't have the data of the entry (or asset). What I would suggest is that we remove it from the structure and then add it to the errors array.

This could still result in runtime errors, but it would not crash on the SDK level. I hope it would also be easier to understand why a certain item is missing from the object structure as it would be included in the error array.

Any thoughts?

**Comment by andersekdahl** [comment]
So for this example:
```
public class MyEntry
{
    public ISomeEntry SomeEntry { get; set; }
    public List<ISomeEntry> SomeEntries { get; set; }
}
```
Would it be that `SomeEntry` is null, and `SomeEntries` becomes an empty list? Both of those sounds fine to me! But if `SomeEntries` can contain nulls then it's a little less appealing, but still better than crashing.

**Comment by Roblinde** [comment]
Yes, that is exactly right. A single interface would be null and a collection would be empty. I will write some more tests and see if I can get a new version out today.

**Comment by Roblinde** [comment]
This should now be fixed in version 5.0.4
